### PR TITLE
procd: remount the rootfs read-only at halt

### PR DESCRIPTION
**Pull request by graysky2**
An ext4 rootfs reports a dirty journal and stale free block/inode counts on every boot, even after an orderly reboot:

      % e2fsck -fvy /dev/nvme0n1p2
      rootfs: recovering journal
      Free blocks count wrong (136398, counted=124525).
      Fix? yes
      Free inodes count wrong (41703, counted=41667).
      Fix? yes
 
`sync()` is not enough, as a journalling filesystem only commits its superblock on remount or unmount. The "umount -a -r" in the umount init script does try a read-only remount, but it runs while procd and the daemons it supervises are still alive, so it fails with EBUSY as long as any of them holds a file open for writing.

The only point in the shutdown where the root is idle is after procd's final SIGKILL, so have procd remount it read-only there, right before it asks the kernel to reboot.

While there, switch procd's log channel to the console on the way into STATE_HALT. logd is long gone by then, so a failed remount would otherwise go unseen.

### CODE REVIEW

**Comment by graysky2** [comment]
https://github.com/openwrt/openwrt/pull/21478 fixes a related problem. Taking together these two pull requests give a cleanly written file system.

**Comment by champtar** [comment]
You can have a look at systemd-shutdown to see if you've handled all corner cases https://github.com/systemd/systemd/blob/main/src/shutdown/shutdown.c
The remount ro might be something you want to do even later than `STOP=90`, maybe even from procd / pid 1 so really nothing is supposed to be running anymore

File: package/base-files/files/etc/init.d/umount [hunk 1]
```diff
@@ -7,7 +7,81 @@ restart() {
...
+# holding a file open for writing outside /tmp and the virtual
+# filesystems. Such descriptors make the read-only remount of a block
+# filesystem fail, leaving a dirty journal for the next boot's fsck.
+disk_writers() {
+	local fd pid target flags
+
+	for fd in /proc/[0-9]*/fd/*; do
+		pid="${fd#/proc/}"
+		pid="${pid%%/*}"
+		[ "$pid" = "1" -o "$pid" = "$$" ] && continue
```
Comment (champtar) [inline package/base-files/files/etc/init.d/umount]: You should log if this script or pid 1 has a RW fd open

File: package/base-files/files/etc/init.d/umount [hunk 2]
```diff
@@ -7,7 +7,81 @@ restart() {
...
+
+		target="$(readlink "$fd" 2> /dev/null)"
+		case "$target" in
+		/dev/* | /proc/* | /sys/* | /tmp/* | /memfd:*) continue ;;
+		/*) ;;
+		*) continue ;;
```
Comment (openwrt-ai) [inline package/base-files/files/etc/init.d/umount:53]: This filter matches a write fd to *any* real file path, not only paths on the root filesystem. An open-for-write fd only blocks `mount -o remount,ro /` when the file lives on the root superblock; a process writing to a file on a separate rw block mount (e.g. a data/USB partition under `/mnt` on an x86 ext4 build — exactly the target of this change) does not block the `/` remount, yet it gets reported under the "processes still writing to disk" message and SIGKILLed as if it were the cause. Is that intended? If not, consider restricting the set to descriptors whose target actually resides on `/` (same device/mount as root), or at least rewording the console message so it doesn't attribute unrelated writers to the failed root remount.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by openwrt-ai — COMMENTED** [review]
Reviewed 1 new commit. The commit message matches the diff and the shell logic (root remount detection, O_ACCMODE last-octal-digit test, remount/kill retry loop) is sound. One scoping observation left inline on `disk_writers`; the pid-1/self logging, STOP=90 timing, and systemd corner-case points already raised in review are not duplicated here.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by champtar** [comment]
There is a lot running at `STOP=90`, I would keep the sync at 90, and move the umount at 97 so it's before mdadm, and see if it's enough.

**Comment by champtar** [comment]
Also are you sure umount silently fails ? https://github.com/mirror/busybox/blob/371fe9f71d445d18be28c82a2a6d82115c8af19d/util-linux/umount.c#L205

**Comment by graysky2** [comment]
@champtar - Thanks very much for the review and the thoughtful feedback. Plz see latest push.

File: package/base-files/files/etc/init.d/umount (hunk 1 above)
Comment (graysky2) [reply package/base-files/files/etc/init.d/umount]: done

**Review by openwrt-ai — COMMENTED** [review]
Reviewed 1 new commit; no new issues found.

The prior review's scoping concern on `disk_writers` (write-fds on any filesystem, not only the root superblock) is now addressed: descriptors are matched by resolving each fd's `mnt_id` to its `major:minor` via `/proc/self/mountinfo` and comparing against the root device, so unrelated writers on other block mounts are no longer reported or killed. The new `swapoff -a`, STOP=90→97 move (mdadm STOP=98 confirmed in-tree), and pid-1/self "pinned" logging all look correct. Commit message matches the diff.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by openwrt-ai — COMMENTED** [review]
Since the last review (`1e3f6c9`) the branch was rebased onto current `main` and the commit message was rewritten; `package/base-files/files/etc/init.d/umount` is byte-identical to the previously reviewed revision, so no code findings are repeated here.

## Commit checks

- `ecf19ca` "base-files: remount rootfs RO on shutdown" — the reworded body reintroduces a claim the previous revision had already corrected. It now says "`umount -a -r` fails silently if processes still hold files open for writing", but busybox's `umount` does print a diagnostic on the fallback read-only remount — [`bb_error_msg("can't remount %s read-only", ...)` at util-linux/umount.c:205](https://github.com/mirror/busybox/blob/371fe9f71d445d18be28c82a2a6d82115c8af19d/util-linux/umount.c#L205), the exact line [@champtar raised](https://github.com/openwrt/openwrt/pull/24053#issuecomment-4878550508). The `1e3f6c9` body had the accurate formulation ("busybox does print `can't remount ... read-only`, but nothing acts on the failure and the message is rarely visible"), which is both true and a stronger justification for the change. Suggest restoring that wording rather than "fails silently".

- `ecf19ca` "base-files: remount rootfs RO on shutdown" — nit: the reword also dropped the statement that only descriptors on the root superblock are considered. That scoping is the substantive change made in response to the first review round (`disk_writers` resolves each fd's `mnt_id` to its `major:minor` via `/proc/self/mountinfo` and compares it against the root device), and it is what keeps writers on unrelated rw block mounts from being reported and SIGKILLed. The bullet "Identifies write-holders if the remount fails" now reads broader than what the code does; worth keeping the "writers on other filesystems are left alone" sentence.

For the record, the two in-tree facts the message asserts still check out on the rebased base: `mdadm.init` is `STOP=98` and `sysfixtime` is `STOP=90`, so `STOP=97` sits where the message says it does. The formality bot's `PKG_RELEASE` warning looks like a false positive here — `package/base-files/Makefile` uses `PKG_RELEASE:=$(COMMITCOUNT)`, so it bumps on its own.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by openwrt-ai — COMMENTED** [review]
Reviewed 1 new commit; no new issues found.

Since the last review (`ecf19ca`) the branch was rebased onto current `main` with no content change: `package/base-files/files/etc/init.d/umount` is byte-identical (both revisions point at blob `02ba3e6a38ac8dd868b0202053c730fb75b7e185`) and the commit message is unchanged as well. So there is nothing new to review in the diff itself, and no code findings are repeated here.

The in-tree facts the message asserts still hold on the rebased base `c0262ed`: `package/base-files/files/etc/init.d/sysfixtime` is `STOP=90` and `package/utils/mdadm/files/mdadm.init` is `STOP=98`, so the move to `STOP=97` still lands where the message says it does. CI on `4b3e859` is green across all check runs.

## Commit checks

- `4b3e859` "base-files: remount rootfs RO on shutdown" — the two commit-message points raised against `ecf19ca` carry over unchanged, since the reword was not applied before the rebase. Details are in the [previous review](https://github.com/openwrt/openwrt/pull/24053#pullrequestreview-4948223750) rather than restated here: (1) the body still says "`umount -a -r` fails silently", which busybox contradicts — it prints `can't remount %s read-only`; the accurate `1e3f6c9` formulation ("nothing acts on the failure and the message is rarely visible") is the stronger justification anyway; (2) nit: the body still omits that only descriptors on the root superblock are considered, which is the substantive fix made in the first review round and the reason writers on unrelated rw block mounts are not SIGKILLed.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by geldot** [comment]
Yeah, it's going to have to be done after procd's final SIGKILL.

Possible approach: Invoke `SHUTDOWN=1 /sbin/mount_root stop` in procd's `STATE_HALT` just after SIGKILL.

`stop()` is currently a stub but it has the context needed to do the right unmount/remount dance for whatever rootfs is in use.

Update: The procd sync()/sleep(1) calls also look like they are the wrong way round in the STATE_HALT case. Normally you want kill()->sleep()-sync(), whereas currently it it's kill()->sync()->sleep() which results in loss of data flushed during the teardown AFAICT. I'd investigate there before doing workarounds

**Review by openwrt-ai — COMMENTED** [review]
Reviewed 1 new commit; no new issues found.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by graysky2** [comment]
@geldot - I redesigned this based on your feedback. What do you think? Works on my x86/64 box.

**Comment by geldot** [comment]
> @geldot - I redesigned this based on your feedback. What do you think? Works on my x86/64 box.

The lifecycle is basically correct, but the patch adds a lot of code and hard-coded assumptions which are best avoided.

I suspect there's a much bigger win and simplification to be had.

Best case: If you eliminate userspace fs teardown and leave as much as possible to procd's SHUTDOWN, there will likely be enough state that the high-level fstools block API will still be mostly usable. It may just need tweaks to get you across the finish line.

Even if that doesn't work, there's still a lot of lower-level reusable infrastructure in the fstools internals that you could use to lighten your current approach.

So those are the two approaches I'd investigate next to slim down your impl.

File: package/system/fstools/patches/0001-mount_root-close-the-root-filesystem-out-on-shutdown.patch [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
++
++	/* a root that only ever lived in RAM has nothing to close out */
++	if (sfs.f_type == TMPFS_MAGIC || sfs.f_type == RAMFS_MAGIC)
++		return 0;
++
++	if (sfs.f_type != OVERLAYFS_MAGIC)
++		return remount_ro("/");
++
++	/* quiescing the overlay flushes the filesystem underneath it */
++	remount_ro("/");
```
Comment (openwrt-ai) [inline package/system/fstools/patches/0001-mount_root-close-the-root-filesystem-out-on-shutdown.patch]: The result is dropped here, while the identical call three lines above is propagated, so `stop()` still exits 0 on an overlay root whose `/` could not be quiesced. procd's `WEXITSTATUS(status)` check and its `failed to close the rootfs` message therefore never fire on exactly the devices that use an overlay. Fold both remounts into the return value.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: package/system/procd/patches/0001-state-close-the-root-filesystem-out-before-halting.patch [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
++
++	pid = fork();
++	if (pid < 0) {
++		ERROR("failed to fork mount_root: %m\n");
++		goto out;
++	}
++
++	if (!pid) {
++		setenv("SHUTDOWN", "1", 1);
++		execl("/sbin/mount_root", "mount_root", "stop", NULL);
```
Comment (openwrt-ai) [inline package/system/procd/patches/0001-state-close-the-root-filesystem-out-before-halting.patch]: nit: procd has no dependency on fstools — [`Package/procd/Default` DEPENDS](https://github.com/openwrt/openwrt/blob/c66025a5aefdbc5fd10a7a997c98f13adef27104/package/system/procd/Makefile#L43-L44) — so on an image built without it this exec fails and every halt logs both `failed to execute mount_root` and `failed to close the rootfs`. Gate the fork on `access("/sbin/mount_root", X_OK)` so a missing helper is not reported as a failed close.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by openwrt-ai — COMMENTED** [review]
## Commit checks

- `c66025a` "procd: fstools: remount rootfs read-only at halt" — the commit also rewrites `package/base-files/files/etc/init.d/umount` to swap off file-backed swap; neither the subject prefix nor the body mentions base-files at all. Split it into its own `base-files: ...` commit, or name the change in both subject and body.
- `c66025a` — nit: body typos: "cloase the root out" → "close", "procd's final final SIGKILL" → one "final", "that will fails with EBUSY" → "fail".

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by graysky2** [comment]
@geldot - better thanks for the suggestion

File: package/system/procd/patches/0001-state-remount-the-root-filesystem-read-only-before-halting.patch
Comment (openwrt-ai) [inline package/system/procd/patches/0001-state-remount-the-root-filesystem-read-only-before-halting.patch:1]: procd is built from [`$(PROJECT_GIT)/project/procd.git` at a pinned `PKG_SOURCE_VERSION`](https://github.com/openwrt/openwrt/blob/83d40d91ec97b6c8226f087cc7bf0590300a2cb9/package/system/procd/Makefile#L13-L17); of the 37 in-tree packages sourced that way, only `package/network/utils/iwinfo` carries a local patch — the others land the change in the upstream project and bump `PKG_SOURCE_VERSION`/`PKG_SOURCE_DATE`/`PKG_MIRROR_HASH` here. Submit this to procd.git and bump the source pin instead of carrying an in-tree patch (which also makes `PKG_RELEASE:=2` unnecessary).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: package/system/procd/patches/0001-state-remount-the-root-filesystem-read-only-before-halting.patch [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
++		/*
++		 * sync() writes the data back, but does not leave the root
++		 * filesystem clean: a journalling filesystem only commits its
++		 * superblock on remount or unmount, and the remount is only
++		 * possible now that no process is left to hold a file on it
++		 * open for writing. The root of a container is the host's.
++		 */
++		if (!is_container() &&
++		    mount(NULL, "/", NULL, MS_REMOUNT | MS_RDONLY, NULL))
++			ERROR("failed to remount / read-only: %m\n");
```
Comment (openwrt-ai) [inline package/system/procd/patches/0001-state-remount-the-root-filesystem-read-only-before-halting.patch:68]: `ERROR()` now resolves through the `ULOG_STDIO` channel installed a few lines above, i.e. `vfprintf(stderr, ...)`, and musl's `printf` does not implement the glibc `%m` extension — only its `syslog()` substitutes it — so on a default OpenWrt build the errno text is dropped from exactly the message this patch adds the channel switch for. Format `strerror(errno)` explicitly here (state.c includes neither `<errno.h>` nor `<string.h>` today, so add them alongside `<sys/mount.h>`).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by openwrt-ai — COMMENTED** [review]
Reviewed 1 new commit.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: package/system/procd/patches/0001-state-remount-the-root-filesystem-read-only-before-halting.patch (hunk 5 above)
Comment (graysky2) [reply package/system/procd/patches/0001-state-remount-the-root-filesystem-read-only-before-halting.patch:68]: %m via ERROR()/vfprintf: not a bug. musl's vfprintf (src/stdio/vfprintf.c) does implement %m via strerror(errno) - verified against musl v1.2.6, the version this tree's toolchain pins. procd already relies on %m through ERROR() in 100+ other call sites, including two others already in state.c, so this call is consistent with existing usage. No code change needed here.

**Comment by graysky2** [comment]
I'm going to close this since I submitted it against procd: https://github.com/openwrt/procd/pull/43
